AEPNotificationContent/Sources/AEPNotificationViewController.swift

```diff
@@ -14,48 +14,93 @@ import UIKit
 import UserNotifications
 import UserNotificationsUI
 
-open class AEPNotificationViewController: UIViewController, UNNotificationContentExtension {
+open class AEPNotificationViewController: UIViewController, UNNotificationContentExtension, TemplateControllerDelegate {
+    var notification: UNNotification?
+
+    // MARK: - UNNotificationContentExtension delegate methods
+
     /// This method is called in the main thread of the notification content app extension
     public func didReceive(_ notification: UNNotification) {
-        guard let payload = notification.request.content.userInfo as? [String: AnyObject],
-              let template = payload[AEPNotificationContentConstants.PayloadKey.TEMPLATE_TYPE] as? String
+        self.notification = notification
+        let notificationContent = notification.request.content
+
+        // Extract the template type from the notification content
+        guard let notificationUserInfo = notification.request.content.userInfo as? [String: AnyObject],
+              let templateType = notificationUserInfo[AEPNotificationContentConstants.PayloadKey.TEMPLATE_TYPE] as? String,
+              let payload = createPayload(for: templateType, from: notificationContent)
         else {
-            displayFallbackTemplate(notification)
+            displayFallbackTemplate(notificationContent)
             return
         }
 
-        switch template {
-        case AEPNotificationContentConstants.PayloadKey.TemplateType.BASIC:
-            if let payload = BasicPayload(from: notification.request.content) {
-                let basicTemplateController = BasicTemplateController(payload)
-                view.addSubview(basicTemplateController.view)
-                preferredContentSize.height = basicTemplateController.preferredContentSize.height
-            } else {
-                displayFallbackTemplate(notification)
-            }
-        case AEPNotificationContentConstants.PayloadKey.TemplateType.CAROUSEL:
-            print("//TODO: Create and display Carousel viewController")
-        case AEPNotificationContentConstants.PayloadKey.TemplateType.TIMER:
-            print("//TODO: Create and display Timer viewController")
-        default:
-            displayFallbackTemplate(notification)
+        // Create a controller based on the template type
+        let controller: UIViewController?
+        if let basicPayload = payload as? BasicPayload {
+            controller = BasicTemplateController(withPayload: basicPayload, delegate: self)
+        } else { controller = nil }
+
+        // If a controller was created, add it as a child and its view to the hierarchy
+        if let controller = controller {
+            addChild(controller)
+            view.addSubview(controller.view)
+        } else {
+            displayFallbackTemplate(notificationContent)
```

Comment:
can add this to the above condition? consolidate the two?

Reply:
when we implement more templates the code here will look like
```
        let controller: UIViewController?
         if let basicPayload = payload as? BasicPayload {
             controller = BasicTemplateController(withPayload: basicPayload, delegate: self)
         } else if let carouselPayload = payload as? CarouselPayload {
             controller = CarouselTemplateController(withPayload: carouselPayload, delegate: self)
         } else if let timerPayload = payload as? TimerPayload {
             controller = TimerTemplateController(withPayload: timerPayload, delegate: self)
         } else {controller = nil}

        // If a controller was created, add it as a child and its view to the hierarchy
        if let controller = controller {
            addChild(controller)
            view.addSubview(controller.view)
        } else {
            displayFallbackTemplate(notificationContent)
        }
```

I will have to assign value to the controller depending on the type of payload..

Hence I split them in two condition
1. For Assiging value to controller
2. Usage